src/osd/PGBackend.cc

```diff
@@ -916,11 +921,21 @@ map<pg_shard_t, ScrubMap *>::const_iterator
     if (shard_info.errors)
       goto out;
 
+    // XXX: Do I want replicated only?
+    if (parent->get_pool().is_replicated() && cct->_conf->osd_distrust_data_digest) {
+      // This is a boost::optional<bool> so see if option set AND it has the value true
+      // We give priority to a replica where the ObjectStore like BlueStore has builtin checksum
+      if (j->second->has_builtin_csum && j->second->has_builtin_csum == true) {
```

Comment:
All these are true when has_builtin_csum optional set to true:

1. has_builtin_csum && *has_builtin_csum
1. has_builtin_csum && has_builtin_csum == true
1. has_builtin_csum == true

The 3rd one would work, but since it is optional the 2nd one might be clearer.

*has_builtin_csum alone would assert if the value of the optional isn't set.

Reply:
Just looking at the code, the left side ``j->second->has_builtin_csum`` and right side ``j->second->has_builtin_csum == true`` look identical: the left is evaluated in a boolean context so it's no different than the right. I think it should be
``j->second->has_builtin_csum && *j->second->has_builtin_csum == true``, so that we only look inside the optional if it has a value.

Reply:
```
dzafman@ubuntu16:~/ceph/build$ more test.cc
#include <boost/optional.hpp>
#include <iostream>
#include "boost/none.hpp"

main()
{
  boost::optional<bool> oEmpty(boost::none), oTrue(true), oFalse(false);

  if (oEmpty && *oEmpty)
    std::cout << "oEmpty && *oEmpty" << std::endl;
  if (oFalse && *oFalse)
    std::cout << "oFalse && *oFalse" << std::endl;
  if (oTrue && *oTrue)
    std::cout << "oTrue && *oTrue" << std::endl;
  if (oEmpty && oEmpty == true)
    std::cout << "oEmpty && oEmpty == true" << std::endl;
  if (oFalse && oFalse == true)
    std::cout << "oFalse && oFalse == true" << std::endl;
  if (oTrue && oTrue == true)
    std::cout << "oTrue && oTrue == true" << std::endl;
  if (oEmpty == true)
    std::cout << "oEmpty == true" << std::endl;
  if (oFalse == true)
    std::cout << "oFalse == true" << std::endl;
  if (oTrue == true)
    std::cout << "oTrue == true" << std::endl;
  exit(0);
}
dzafman@ubuntu16:~/ceph/build$ ./test
oTrue && *oTrue
oTrue && oTrue == true
oTrue == true
````

Reply:
Gotcha.  Must be the difference between operator== and operator bool().  Anyway, LGTM!